utils/math-utils/src/main/scala/pl/touk/nussknacker/engine/util/MathUtils.scala

```diff
@@ -191,15 +197,34 @@ trait MathUtils {
   }
 
   protected trait SameNumericTypeHandler[R] {
-    def onBytes(n1: Byte, n2: Byte): R
-    def onShorts(n1: Short, n2: Short): R
-    def onInts(n1: Int, n2: Int): R
-    def onLongs(n1: Long, n2: Long): R
+    def onBytes(n1: java.lang.Byte, n2: java.lang.Byte): R
+    def onShorts(n1: java.lang.Short, n2: java.lang.Short): R
+    def onInts(n1: java.lang.Integer, n2: java.lang.Integer): R
+    def onLongs(n1: java.lang.Long, n2: java.lang.Long): R
     def onBigIntegers(n1: java.math.BigInteger, n2: java.math.BigInteger): R
-    def onFloats(n1: Float, n2: Float): R
-    def onDoubles(n1: Double, n2: Double): R
+    def onFloats(n1: java.lang.Float, n2: java.lang.Float): R
+    def onDoubles(n1: java.lang.Double, n2: java.lang.Double): R
     def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): R
   }
+
+  protected trait SameNumericTypeHandlerReturningNumber extends SameNumericTypeHandler[Number] {
+    override def onBytes(n1: lang.Byte, n2: lang.Byte): java.lang.Byte
+    override def onShorts(n1: lang.Short, n2: lang.Short): java.lang.Short
+    override def onInts(n1: java.lang.Integer, n2: java.lang.Integer): java.lang.Integer
+    override def onLongs(n1: java.lang.Long, n2: java.lang.Long): java.lang.Long
+    override def onBigIntegers(n1: java.math.BigInteger, n2: java.math.BigInteger): java.math.BigInteger
+    override def onFloats(n1: java.lang.Float, n2: java.lang.Float): java.lang.Float
+    override def onDoubles(n1: java.lang.Double, n2: java.lang.Double): java.lang.Double
+    override def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): java.math.BigDecimal
+  }
+
+  protected trait SameNumericTypeHandlerForMathOp extends SameNumericTypeHandlerReturningNumber {
```

Comment:
Maybe `SameNumericTypeHandlerForPromotingMathOp`? Because it is not good choice for all math ops.